Chatto/sources/ChatController/ChatMessages/ChatMessageCollectionAdapter.swift

```diff
@@ -0,0 +1,702 @@
+                        completion: completion
+                    )
+                }
+                return
+            }
+            // ... if they are still invalid the only thing we can do is a reloadData
+            let mustDoReloadData = !visibleCellsAreValid // Only way to recover from this inconsistent state
+            usesBatchUpdates = !wantsReloadData && !mustDoReloadData
+        }
+
+        let scrollAction: ScrollAction
+        do { // Scroll action
+            if updateType != .pagination && collectionView.isScrolledAtBottom() {
+                scrollAction = .scrollToBottom
+            } else {
+                let (oldReferenceIndexPath, newReferenceIndexPath) = self.referenceIndexPathsToRestoreScrollPositionOnUpdate(itemsBeforeUpdate: self.chatItemCompanionCollection, changes: changes)
+                let oldRect = self.rectAtIndexPath(oldReferenceIndexPath)
+                scrollAction = .preservePosition(
+                    rectForReferenceIndexPathBeforeUpdate: oldRect,
+                    referenceIndexPathAfterUpdate: newReferenceIndexPath
+                )
+            }
+        }
+
+        let myCompletion: () -> Void
+        do { // Completion
+            var myCompletionExecuted = false
+            myCompletion = {
+                if myCompletionExecuted { return }
+                myCompletionExecuted = true
+                completion()
+            }
+        }
+
+        if usesBatchUpdates {
+            UIView.animate(withDuration: self.configuration.updatesAnimationDuration, animations: { () -> Void in
+                self.unfinishedBatchUpdatesCount += 1
+                collectionView.performBatchUpdates({
+                    updateModelClosure()
+                    self.updateVisibleCells(changes) // For instance, to support removal of tails
+
+                    collectionView.deleteItems(at: Array(changes.deletedIndexPaths))
+                    collectionView.insertItems(at: Array(changes.insertedIndexPaths))
+                    for move in changes.movedIndexPaths {
+                        collectionView.moveItem(at: move.indexPathOld, to: move.indexPathNew)
+                    }
+                }, completion: { [weak self] _ in
+                    defer { myCompletion() }
+                    guard let sSelf = self else { return }
+                    sSelf.unfinishedBatchUpdatesCount -= 1
+                    if sSelf.unfinishedBatchUpdatesCount == 0, let onAllBatchUpdatesFinished = self?.onAllBatchUpdatesFinished {
+                        DispatchQueue.main.async(execute: onAllBatchUpdatesFinished)
+                    }
+                })
+            })
+        } else {
+            self.visibleCells = [:]
+            updateModelClosure()
+            collectionView.reloadData()
+            collectionView.collectionViewLayout.prepare()
+        }
+
+        collectionView.setNeedsLayout()
+        collectionView.layoutIfNeeded()
```

Comment:
Is it possible avoid forcing layout here?

Reply:
This was causing some issues to calculate the new content size that would impact the next method to scroll to the bottom :/ This was one of the ways I could find out out to fix it